clients/go/compat.go

```diff
@@ -0,0 +1,306 @@
+	req.Header.Set("Content-Type", "application/json")
+
+	client := http.Client{}
+	res, err := client.Do(req)
+
+	if err != nil {
+		return nil, err
+	}
+
+	decoder := json.NewDecoder(res.Body)
+	rpcRes := new(Response)
+	if err := decoder.Decode(&rpcRes); err != nil {
+		return nil, err
+	}
+
+	if err := res.Body.Close(); err != nil {
+		return nil, err
+	}
+
+	var pubKey oasis.CallDataPublicKey
+	json.Unmarshal(rpcRes.Result, &pubKey)
+
+	return pubKey.PublicKey, nil
+}
+
+func NewWrappedBackend(backend bind.ContractBackend, transactOpts *bind.TransactOpts, chainID *big.Int, privateKey *ecdsa.PrivateKey, signerFn func(digest [32]byte) ([]byte, error)) (*WrappedBackend, error) {
+	runtimePublicKey, err := GetRuntimePublicKey()
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &WrappedBackend{
+		Backend:          backend,
+		ChainID:          *chainID,
+		Key:              *privateKey,
+		RuntimePublicKey: runtimePublicKey,
+		TransactOpts:     *transactOpts,
+		Signer:           NewSigner(signerFn),
+	}, nil
+}
+
+func (w WrappedSigner) Sign(digest [32]byte) ([]byte, error) {
+	return w.SignFn(digest)
+}
+
+func (b WrappedBackend) SendTransaction(ctx context.Context, tx *types.Transaction) error {
+	header, err := b.Backend.HeaderByNumber(ctx, nil)
+	if err != nil {
+		return err
+	}
+
+	blockHash := header.Hash()
+	leash := NewLeash(header.Nonce.Uint64(), header.Number.Uint64(), blockHash[:], DefaultBlockRange)
+	publicKey, _ := tweetnacl.ScalarMultBase(crypto.FromECDSA(&b.Key))
+	keypair := tweetnacl.KeyPair{
+		PublicKey: publicKey,
+		SecretKey: crypto.FromECDSA(&b.Key),
+	}
+	cipher, _ := NewX255919DeoxysIICipher(keypair, b.RuntimePublicKey)
+	dataPack, _ := NewDataPack(b.Signer, tx.ChainId().Uint64(), b.TransactOpts.From[:], addressToByte(tx.To()), tx.Gas(), tx.GasPrice(), tx.Value(), tx.Data(), leash)
+
+	legacyTx := &types.LegacyTx{
+		To:       tx.To(),
+		Nonce:    tx.Nonce(),
+		GasPrice: tx.GasPrice(),
+		Gas:      DefaultGasLimit,
+		Value:    tx.Value(),
+		Data:     dataPack.EncryptEncode(cipher),
+	}
+
+	baseTx := *types.NewTx(legacyTx)
+
+	signer := types.LatestSignerForChainID(tx.ChainId())
+	signature, _ := crypto.Sign(signer.Hash(&baseTx).Bytes(), &b.Key)
+	signedTx, _ := baseTx.WithSignature(signer, signature)
+
+	return b.Backend.SendTransaction(ctx, signedTx)
+}
+
+func (b WrappedBackend) CallContract(ctx context.Context, call ethereum.CallMsg, blockNumber *big.Int) ([]byte, error) {
+	header, err := b.Backend.HeaderByNumber(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	blockHash := header.Hash()
+	leash := NewLeash(header.Nonce.Uint64(), header.Number.Uint64(), blockHash[:], DefaultBlockRange)
+
+	publicKey, _ := tweetnacl.ScalarMultBase(crypto.FromECDSA(&b.Key))
+	keypair := tweetnacl.KeyPair{
+		PublicKey: publicKey,
+		SecretKey: crypto.FromECDSA(&b.Key),
+	}
+	cipher, _ := NewX255919DeoxysIICipher(keypair, b.RuntimePublicKey)
```

Comment:
can you cache this in the wrapped backend? the cipher has a bit of setup cost that could be avoided.